src/utils/lock.js

```diff
@@ -49,6 +50,6 @@ module.exports = class Lock {
   async release() {
     this[PRIVATE.LOCKED] = false
     const waitingForLock = Array.from(this[PRIVATE.WAITING])
-    return Promise.all(waitingForLock.map(acquireLock => acquireLock()))
+    return promiseAllInBatches({ batchSize: 20 }, waitingForLock)
```

Comment:
I understand what it's doing, but I'm trying to work through in my head why this would have an impact.

Say we have 1000 things that all want to acquire a lock that is currently locked. At that moment, we have 1000 functions in the wait list. Once the lock is released, we would call those 1000 functions. One of them would manage to acquire the lock and delete itself from the wait list. The other ones would do literally nothing. We now have 999 functions in the wait list. And so it repeats until all of them have either timed out or had a chance to acquire the lock.

With the new implementation, the only difference is that we only call the first 20 waiting functions instead of all 1000. This does save us some calls, but since we know only one of them can acquire the lock anyway, why don't we just pop the first one instead of calling 20 of them?

Reply:
The difference is that we have 1000 `async` functions, which means we are creating a lot of promises and increasing memory and CPU usage. The system ends up in a huge pause and all the promises timeout once the event loop reaches the time queue, this change makes the system more stable even if the user is generating a lot of promises in userland code.

Reply:
Sure, that's true. Not to mention that when the lock releases the next time, you'll generate 998 promises, then 997, etc. Which is quite a large series. But the question remains, why even have 20 of them competing against each other? Why not just take a single one?

Reply:
you have a point

Reply:
@Nevon a3552cb is a lot more efficient now, we were creating yet another array with `Array.from` to convert the `Set`, now I am just using the iterator.